### PR TITLE
feat(nep5): add nep5 balance/transfer models and storage

### PR DESCRIPTION
**Pull request by spencercorwin**
### Description of the Change

Adds storage for NEP5 transfers and balances. Add RPC methods for getting NEP5 balances/transfers.

### Test Plan

Test in final e2e testing.

### Alternate Designs

None.

### Benefits

Closer to Neo3. Can now track NEP5 balances/transfers.

### Possible Drawbacks

None.

### Applicable Issues

#2156 
#2009 
#2142 
#1882

### CODE REVIEW

File: packages/neo-one-client-common/src/models/nep5/Nep5TransferModel.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { BinaryWriter, createSerializeWire, SerializableWire, SerializeWire, UInt160, UInt256 } from '../..';
```
Comment (danwbyrne) [inline packages/neo-one-client-common/src/models/nep5/Nep5TransferModel.ts:1]: fix this import

File: packages/neo-one-client-common/src/models/nep5/Nep5TransferKeyModel.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import { BN } from 'bn.js';
+import { BinaryWriter, createSerializeWire, SerializableWire, SerializeWire, UInt160 } from '../..';
```
Comment (danwbyrne) [inline packages/neo-one-client-common/src/models/nep5/Nep5TransferKeyModel.ts:2]: fix this import

File: packages/neo-one-client-common/src/models/nep5/Nep5BalanceModel.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { BinaryWriter, createSerializeWire, SerializableWire, SerializeWire } from '../..';
```
Comment (danwbyrne) [inline packages/neo-one-client-common/src/models/nep5/Nep5BalanceModel.ts:1]: fix this import

File: packages/neo-one-client-common/src/models/nep5/Nep5BalanceKeyModel.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { BinaryWriter, createSerializeWire, SerializableWire, SerializeWire, UInt160 } from '../..';
```
Comment (danwbyrne) [inline packages/neo-one-client-common/src/models/nep5/Nep5BalanceKeyModel.ts:1]: fix this import

File: packages/neo-one-node-blockchain/src/Blockchain.ts [hunk 5]
```diff
@@ -643,4 +665,174 @@ export class Blockchain {
...
+        if (stateItems.length === 0) {
+          return;
+        }
+        if (eventName !== 'Transfer') {
+          return;
+        }
+        if (stateItems.length < 3) {
+          return;
+        }
+        if (!stateItems[0].isNull && stateItems[0].type !== StackItemType.ByteString) {
```
Comment (danwbyrne) [inline packages/neo-one-node-blockchain/src/Blockchain.ts:693]: I think we need to move all of this to another file, very large function and the only blockchain method it needs is `runEngineWrapper`.

File: packages/neo-one-node-core/src/Serializable.ts [hunk 6]
```diff
@@ -5,11 +5,13 @@
...
 // import { StorageChange } from './storageChange';
 import { SerializableWire } from '@neo-one/client-common';
+import { Block } from './Block';
+import { Transaction } from './transaction';
 import { BinaryReader } from './utils';
 
 export { SerializeWire, SerializableWire, createSerializeWire } from '@neo-one/client-common';
 
-export type SerializableContainerType = 'Block' | 'Signers' | 'Transaction';
+export type SerializableContainerType = 'Block' | 'Transaction';
```
Comment (danwbyrne) [inline packages/neo-one-node-core/src/Serializable.ts:14]: this will get changed in a rebase

File: packages/neo-one-node-core/src/utils/utils.ts [hunk 7]
```diff
@@ -201,8 +201,8 @@ const wildCardFromJSON = <T>(json: WildcardContainerJSON, selector: (input: stri
 };
 
 /* crude method but it does what we want it to do */
-const generateSearchRange = (lookupKey: Buffer): StreamOptions => {
-  const asBN = new BN(lookupKey);
+const generateSearchRange = (lookupKey: Buffer): Required<StreamOptions> => {
+  const asBN = new BN(lookupKey, 'le');
```
Comment (danwbyrne) [inline packages/neo-one-node-core/src/utils/utils.ts:205]: will get rebased out

File: packages/neo-one-node-http-rpc/src/setupRPCServer.ts [hunk 8]
```diff
@@ -2,6 +2,7 @@ import { cors, setupServer } from '@neo-one/http';
 import { context, onError as appOnError } from '@neo-one/http-context';
 import { createChild, nodeLogger, rpcLogger } from '@neo-one/logger';
 import { Blockchain, Node } from '@neo-one/node-core';
+import { NativeContainer } from '@neo-one/node-native';
 import { Disposable, Labels } from '@neo-one/utils';
 import * as nodeHttp from 'http';
 import Application from 'koa';
```
Comment (danwbyrne) [inline packages/neo-one-node-http-rpc/src/setupRPCServer.ts:8]: I don't think you need NativeContainer in the RPC server, blockchain already has a reference to native

File: packages/neo-one-node-http-rpc/src/middleware/rpc.ts [hunk 9]
```diff
@@ -1,16 +1,26 @@
...
+  readonly blockchain: Blockchain;
+  readonly node: Node;
+  readonly native: NativeContainer;
+}) => {
   const handler = createHandler({
     blockchain,
     node,
+    native,
     handleGetNEOTrackerURL: async () => {
       const config = await loadConfiguration();
```
Comment (danwbyrne) [inline packages/neo-one-node-http-rpc/src/middleware/rpc.ts:25]: why do we need this

File: packages/neo-one-node-storage-levelup/src/levelUpStorage.ts [hunk 10]
```diff
@@ -8,15 +8,19 @@ import {
...
+  Nep5TransferKey,
+  Storage,
   StorageItem,
   StorageKey,
   TransactionState,
   TrimmedBlock,
 } from '@neo-one/node-core';
 import { keys } from '@neo-one/node-storage-common';
 import { LevelUp } from 'levelup';
+import { Nep5BalanceKey } from '../../neo-one-node-core/src/Nep5BalanceKey';
```
Comment (danwbyrne) [inline packages/neo-one-node-storage-levelup/src/levelUpStorage.ts:23]: fix import

File: packages/neo-one-node-vm/lib/Dispatcher.csproj [hunk 11]
```diff
@@ -26,8 +26,8 @@
   </ItemGroup>
 
   <ItemGroup>
-    <ProjectReference Include="/Users/danielbyrne/neo/src/neo/neo.csproj" />
-    <ProjectReference Include="/Users/danielbyrne/neo-vm/src/neo-vm/neo-vm.csproj" />
+    <ProjectReference Include="/Users/spencercorwin/neo/src/neo/neo.csproj" />
+    <ProjectReference Include="/Users/spencercorwin/neo-vm/src/neo-vm/neo-vm.csproj" />
```
Comment (danwbyrne) [inline packages/neo-one-node-vm/lib/Dispatcher.csproj:30]: sorry chief there can only be ONE of us who has to change this line after every rebase and I already called dibs.

File: packages/neo-one-node/src/startFullNode.ts [hunk 12]
```diff
@@ -130,6 +130,7 @@ export const startFullNode = async ({
       blockchain,
       node,
       options: rpcOptions,
+      native,
```
Comment (danwbyrne) [inline packages/neo-one-node/src/startFullNode.ts:133]: we can talk about this what the plan with this native pass down was.

**Review by danwbyrne — CHANGES_REQUESTED** [review]

File: packages/neo-one-node-http-rpc/src/middleware/rpc.ts (hunk 9 above)
Comment (spencercorwin) [reply packages/neo-one-node-http-rpc/src/middleware/rpc.ts:25]: Two methods in `createHandler()` (`getvalidators`, `getunclaimedgas`) were already trying access `native` without having an actual reference to `native`

File: packages/neo-one-node-blockchain/src/Blockchain.ts (hunk 5 above)
Comment (spencercorwin) [reply packages/neo-one-node-blockchain/src/Blockchain.ts:693]: Also settings, but still easy to do.

**Pull request merged**